⚠️ move loadbalancer opts to struct

apricote opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:

Modify the `v1beta1` CRD and move the loadbalancer fields `OpenStackClusterSpec.{ManagedAPIServerLoadBalancer,APIServerLoadBalancerAdditionalPorts}` into a struct at `OpenStackClusterSpec.APIServerLoadBalancer` to enable adding further customizations. This was previously discussed in #1105.

**Which issue(s) this PR fixes**:
Related to #1105

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

jichenjc commented:
Thanks for the PR, I assume you need update templates/*.yaml (e.g cluster-template.yaml) and maybe e2e settings 
in order to honor the new things introduced?

apricote commented:
I just pushed an update that contains:

- Unit tests for the conversions
- Fixed `OpenStackCluster` manifests in `templates/` and `test/data/infrastructure-openstack`
- updated generated conversions (`make generate`)

jichenjc commented:
/lgtm

thanks for the fix ~

mdbooth commented:
This is a really nice cleanup.

/approve

apricote commented:
/hold cancel

apricote commented:
Looks like a flaky test, was previously successful on the same commit.

/retest

chrischdi commented on api/v1alpha3/conversion.go:
```diff
@@ -117,6 +119,10 @@ func Convert_v1alpha3_OpenStackClusterSpec_To_v1beta1_OpenStackClusterSpec(in *O
 	}
+	out.APIServerLoadBalancer = infrav1.APIServerLoadBalancer{
+		Enabled:         in.ManagedAPIServerLoadBalancer,
+		AdditionalPorts: *(*[]int)(unsafe.Pointer(&in.APIServerLoadBalancerAdditionalPorts)),
```
What's the reason for using unsafe here? couldn't it be as easy as the following?:

```suggestion
		AdditionalPorts: in.APIServerLoadBalancerAdditionalPorts,
```

chrischdi commented on api/v1alpha3/conversion.go:
```diff
@@ -139,6 +145,10 @@ func Convert_v1beta1_OpenStackClusterSpec_To_v1alpha3_OpenStackClusterSpec(in *i
 	}
+
+	out.ManagedAPIServerLoadBalancer = in.APIServerLoadBalancer.Enabled
+	out.APIServerLoadBalancerAdditionalPorts = *(*[]int)(unsafe.Pointer(&in.APIServerLoadBalancer.AdditionalPorts))
```
What's the reason for using unsafe here? couldn't it be as easy as the following?:

```suggestion
	out.APIServerLoadBalancerAdditionalPorts = in.APIServerLoadBalancer.AdditionalPorts
```

apricote replied:
I was too slow with answering, not its merged :(

I copied this from the previous autogenerated conversion.

The code is coming from `conversion-gen`: https://github.com/kubernetes/code-generator/blob/7e38d5771c5636f13555f0058e5936c521d2fe13/cmd/conversion-gen/generators/conversion.go#L929

chrischdi replied:
So this whole part is auto generated? If so we have to keep it, otherwise I would prefer to clean it up in a follow-up PR :-)

apricote replied:
Nope, my changes are not auto-generated, but I did not understand the semantics of the conversion code enough to remove the `unsafe.Pointer` logic provided by conversion-gen.
